jujugui/static/gui/src/app/components/machine-view/machine/machine.js

```diff
@@ -71,13 +73,125 @@ YUI.add('machine-view-machine', function() {
       showConstraints: React.PropTypes.bool,
       type: React.PropTypes.string.isRequired,
-      units: React.PropTypes.object.isRequired
+      units: React.PropTypes.object.isRequired,
+      updateMachineConstraints: React.PropTypes.func,
+      updateMachineSeries: React.PropTypes.func
+    },
+
+    getInitialState: function() {
+      return {
+        constraints: null,
+        showForm: false
+      };
+    },
+
+    /**
+      Toggle the display of the constraints form.
+    */
+    _toggleForm: function() {
+      this.setState({showForm: !this.state.showForm});
+    },
+
+    /**
+      Update the state with the new constraints.
+
+      @param constraints {Object} The new constraints.
+    */
+    _updateConstraints: function(constraints) {
+      this.setState({constraints: constraints});
+    },
+
+    /**
+      Set the new constraints on the machine.
+    */
+    _setConstraints: function() {
+      const constraints = this.state.constraints;
+      const series = constraints.series || null;
+      delete constraints.series;
+      const id = this.props.machine.id;
+      this.props.updateMachineConstraints(id, constraints);
+      this.props.updateMachineSeries(id, series);
+      this._toggleForm();
+    },
+
+    /**
+      Generate the constraints form.
+
+      @returns {Object} the form JSX.
+    */
+    _generateConstraintsForm: function() {
+      if (!this.state.showForm) {
+        return null;
+      }
+      const machine = this.props.machine;
+      const disabled = this.props.acl.isReadOnly();
+      const units = this.props.units.filterByMachine(
+        machine.id, this.props.type === 'machine');
+      const buttons = [{
+        title: 'Cancel',
+        action: this._toggleForm,
+        type: 'base'
+      }, {
+        title: 'Update',
+        action: this._setConstraints,
+        type: 'neutral',
+        disabled: disabled
+      }];
+      return (
+        <div className="add-machine__constraints">
+          <h4 className="add-machine__title">
+            Update constraints
+          </h4>
+          <juju.components.Constraints
+            constraints={this._getConstraints()}
+            currentSeries={machine.series}
+            disabled={disabled}
+            hasUnit={!!units.length}
+            providerType={this.props.providerType}
+            series={this.props.series}
+            valuesChanged={this._updateConstraints} />
+          <juju.components.ButtonRow
+            buttons={buttons}
+            key="buttons" />
+        </div>);
+    },
+
+    /**
+      Get the constraints for the machine.
+
+      @returns {Object} the machine constraints.
+    */
+    _getConstraints: function() {
+      const constraints = this.props.machine.constraints || '';
+      let types = {
+        'cpu-power': null,
+        cores: null,
+        mem: null,
+        'root-disk': null,
+        arch: null
+      };
+      // The constraints are in the format:
```

Comment:
The machine stores the constraints as a string. I've updated the comment to clarify that.

As to why that's the case, I'm not entirely sure.